modules/tinymce/src/plugins/link/test/ts/browser/NormalizeLinkTest.ts

```diff
@@ -0,0 +1,96 @@
+    const endContainer = Hierarchy.follow(scope, range.endPath).getOrDie('Failed to resolve end path');
+    rng.setStart(startContainer.dom, range.startOffset);
+    rng.setEnd(endContainer.dom, range.endOffset);
+    return rng;
+  };
+
+  const test = (html: string, range: BaseRange, expectedLink: Optional<number[]>): void => {
+    const editor = hook.editor();
+    const scope = SugarElement.fromDom(editor.getBody());
+    editor.setContent(html);
+    const resolvedRange = resolveTestRange(scope, range);
+    editor.selection.setRng(resolvedRange);
+    const link = getSelectedLink(editor, editor.selection.getNode());
+    const solvedLink = link ? SugarElement.fromDom(link) : SugarElement.fromHtml('<b><b/>');
+    const path = Hierarchy.path(scope, solvedLink);
+
+    KAssert.eqOptional('Should be expected path', path, expectedLink);
+  };
+
+  describe('Returns a link if the end selection is after the link but before a br', () => {
+    it('TINY-6508: Single link', () => {
+      test('<p><a href="tiny">test</a><br></p>', makeRange([ 0, 0, 0 ], 0, [ 0 ], 1), Optional.some([ 0, 0 ]));
+    });
+
+    it('TINY-6508: Multiple short ended elements', () => {
+      test('<p><a href="tiny">test</a><br><img><br></p>', makeRange([ 0, 0, 0 ], 0, [ 0 ], 3), Optional.some([ 0, 0 ]));
+    });
+
+    it('TINY-6508: Nested link', () => {
+      test('<p><b><a href="tiny">test</a></b><br></p>', makeRange([ 0, 0, 0, 0 ], 0, [ 0 ], 1), Optional.some([ 0, 0, 0 ]));
+    });
+
+    it('TINY-6508: Single link with one level nested child', () => {
+      test('<p><a href="tiny"><strong>word</strong>other</a><br></p>', makeRange([ 0, 0, 0, 0 ], 0, [ 0 ], 1), Optional.some([ 0, 0 ]));
+    });
+
+    it('TINY-6508: Single link with two level nested child', () => {
+      test('<p><a href="tiny"><strong><em>word<em/></strong>other</a><br></p>', makeRange([ 0, 0, 0, 0, 0 ], 0, [ 0 ], 1), Optional.some([ 0, 0 ]));
+    });
+
+    it('TINY-6508: Nested link at the end of the line', () => {
+      test('<p><em>test</em><b><a href="tiny">test</a></b><br></p>', makeRange([ 0, 1, 0 ], 0, [ 0 ], 2), Optional.some([ 0, 1, 0 ]));
+    });
+
+    it('TINY-6508: Multiple links', () => {
+      test('<p><a href="tiny">link1</a>content<a href="tiny">link2</a><br></p>', makeRange([ 0, 0, 0 ], 0, [ 0 ], 3), Optional.some([ 0, 0 ]));
+    });
+
+    it('TINY-6508: Simple link with nested Img', () => {
+      test('<p><a href="tiny"><img src="image.jpg"></a>text<br></p>', makeRange([ 0, 0, 0 ], 0, [ 0 ], 2), Optional.some([ 0, 0 ]));
+    });
+  });
+
+  it('TINY-6508: Return a link if the selection starts in the link and end in the link text', () => {
+    test('<p><a href="tiny">link</a></p>', makeRange([ 0 ], 0, [ 0, 0, 0 ], 4), Optional.some([ 0, 0 ]));
+  });
+
+  it('TINY-6508: Does not return a link since the caret is after the link', () => {
+    test('<p><a href="tiny">x</a><br></p>', makeCaretRange([ 0 ], 1), Optional.none());
```

Comment:
There is a lot more none cases we could check for here. Like a range where the end if in another paragraph. Where the range is between two BRs at the end.

A few cases, there could be more but I guess the most important ones is the ones that deals with BRs since it's checking for that then it traverses:
`<p><a href="tiny">[test</a><br>]<br></p>`
`<p><a href="tiny">[test</a><br></p><p>]<br></p>`

Reply:
Code updated and new cases added.